src/coreclr/tools/superpmi/superpmi/neardiffer.cpp

```diff
@@ -435,6 +438,103 @@ bool NearDiffer::compareOffsets(
+    //
+    // Here, we try to match this sequence and look it up in the address map.
+    // 
+    // Some version of this logic might apply to ARM as well.
+    //
+    if (targetArch == SPMI_TARGET_ARCHITECTURE_ARM64)
+    {
+        bool movk2 = false, movk3 = false;
+        unsigned reg1_1, reg1_2, reg2_1, reg2_2, reg3_1, reg3_2, reg4_1, reg4_2;
+        unsigned con1_1, con1_2, con2_1, con2_2, con3_1, con3_2, con4_1, con4_2;
+        unsigned shift2_1, shift2_2, shift3_1, shift3_2, shift4_1, shift4_2;
+        UINT32* iaddr1    = (UINT32*)(data->block1 + blockOffset);
+        UINT32* iaddr2    = (UINT32*)(data->block2 + blockOffset);
+        UINT32* iaddr1end = (UINT32*)(data->block1 + data->blocksize1);
+        UINT32* iaddr2end = (UINT32*)(data->block2 + data->blocksize2);
+
+        // We're assuming that a mov/movk isn't the last instruction in the instruction buffer.
+        if ((iaddr1 < iaddr1end) &&
+            (iaddr2 < iaddr2end) &&
+            GetArm64MovConstant(iaddr1, &reg1_1, &con1_1) &&
+            GetArm64MovConstant(iaddr2, &reg1_2, &con1_2) &&
+            (reg1_1 == reg1_2))
+        {
+            if ((iaddr1 + 1 < iaddr1end) &&
+                (iaddr2 + 1 < iaddr2end) &&
+                GetArm64MovkConstant(iaddr1 + 1, &reg2_1, &con2_1, &shift2_1) &&
+                GetArm64MovkConstant(iaddr2 + 1, &reg2_2, &con2_2, &shift2_2) &&
+                (reg2_1 == reg2_2) &&
+                (shift2_1 == shift2_2) &&
+                (shift2_1 == 16))
+            {
+                DWORDLONG addr1 = (DWORDLONG)con1_1 + ((DWORDLONG)con2_1 << 16);
+                DWORDLONG addr2 = (DWORDLONG)con1_2 + ((DWORDLONG)con2_2 << 16);
+
+                // We currently assume the address requires at least 1 'movk' instruction.
+                if ((iaddr1 + 2 < iaddr1end) &&
+                    (iaddr2 + 2 < iaddr2end) &&
+                    GetArm64MovkConstant(iaddr1 + 2, &reg3_1, &con3_1, &shift3_1) &&
+                    GetArm64MovkConstant(iaddr2 + 2, &reg3_2, &con3_2, &shift3_2) &&
+                    (reg3_1 == reg3_2) &&
+                    (shift3_1 == shift3_2) &&
+                    (shift3_1 == 32))
+                {
+                    movk2 = true;
+                    addr1 += (DWORDLONG)con3_1 << 32;
+                    addr2 += (DWORDLONG)con3_2 << 32;
+
+                    if ((iaddr1 + 3 < iaddr1end) &&
+                        (iaddr2 + 3 < iaddr2end) &&
+                        GetArm64MovkConstant(iaddr1 + 3, &reg4_1, &con4_1, &shift4_1) &&
+                        GetArm64MovkConstant(iaddr2 + 3, &reg4_2, &con4_2, &shift4_2) &&
+                        (reg4_1 == reg4_2) &&
+                        (shift4_1 == shift4_2) &&
+                        (shift4_1 == 48))
+                    {
+                        movk3 = true;
+                        addr1 += (DWORDLONG)con4_1 << 48;
+                        addr2 += (DWORDLONG)con4_2 << 48;
+                    }
+                }
```

Comment:
Do I understand right that this only works if both sides end up with the same number of instructions? For example, it is conceivable that one side may omit a `movk` because that section of the constant is all zeros.

Ultimately, I feel like we need to redesign some of the JIT-EE parts to make this robust. E.g. @EgorBo ended up needing to redesign the JIT-EE interface in #76112 in a way that ends up fixing the problem for good, by splitting the addresses that JIT embeds in codegen from the notion of where the values are. This also helps support things in R2R and NAOT.

Can the same sort of fix not be applied to PGO? I already see that `allocPgoInstrumentationBySchema` returns a `profileMemory` pointer. Is anything stopping us from returning the same pointer to both JITs? From my cursory glance, it seems like it would work and it would make things completely deterministic.

Reply:
You're right: it's checking if the instruction stream is exactly the same modulo the constants, you're right that one `movk` could be omitted if it is zero, or perhaps (unlikely) one could have an address in the >48 bit range and one does not.

> Is anything stopping us from returning the same pointer to both JITs?

The fact that two compilations are happening is unknown at this point in the code. In fact, we currently don't even remember we allocated the buffer (except to put it on a list to free).

We would need to record it as a map from ftnHnd => pointer. There's nothing preventing the JIT from calling the `allocPgoInstrumentationBySchema` (or similar API) multiple times and getting multiple different results.

I guess if the MethodContext remembers if it ever returned a pointer for a ftnHnd, then it can look up in the table and return the same pointer.

Reply:
> The fact that two compilations are happening is unknown at this point in the code. In fact, we currently don't even remember we allocated the buffer (except to put it on a list to free).
> 
> We would need to record it as a map from ftnHnd => pointer. There's nothing preventing the JIT from calling the `allocPgoInstrumentationBySchema` (or similar API) multiple times and getting multiple different results.
> 
> I guess if the MethodContext remembers if it ever returned a pointer for a ftnHnd, then it can look up in the table and return the same pointer.

That's true, it would need an SPMI update (and by extension, JIT-EE GUID update) to save `profileMemory` during recording. I think then keying it by `ftnHnd` would make sense, it is the relationship in the VM today because the VM needs to be able to hand the same pointer back for multiple versions of the same method.

Another benefit is that there are other things that can affect the codegen. Constant CSE could potentially affect the codegen depending on the addresses too. So it would be best if we get the same exact pointer that was returned at runtime to make sure we are seeing the exact same codegen.

Reply:
> By "runtime" do you mean during collection time? I don't think that's possible in general because the JIT might want to write to or in some cases read from that address, and we can't guarantee the same pointer at replay as at collection time.

This is what I was saying previously. I think we should be trying to design the JIT-EE interface so that the JIT does not need to read directly from handles it is using for codegen purposes. This is necessary not just for SPMI, but usually also for R2R/NAOT where the handles are not actually memory addresses. @EgorBo is moving static field addresses to this model in #76112.

I think it is already the case with `allocPgoInstrumentationBySchema` today. From a cursory glance the JIT does not read or write to the returned `profileMemory` pointer. It is purely used for codegen purposes.

If the JIT wants to start reading or writing this pointer I think it should be accomplished with changes to the JIT-EE interface. That way we could also support instrumentation during prejit, for example. In those cases pointers like these are presumably going to be some opaque handle that cannot just be dereferenced.